### PR TITLE
gh-117613: Argument Clinic: disallow `defining_class` parameter at module level

### PR DESCRIPTION
**Pull request by neonene**
This PR follows #117781.

cc @vstinner @erlend-aasland

* Issue: gh-117613

### CODE REVIEW

File: Tools/clinic/libclinic/dsl_parser.py [hunk 1]
```diff
@@ -1102,6 +1102,8 @@ def bad_node(self, node: ast.AST) -> None:
                     fail("A 'defining_class' parameter cannot have a default value.")
                 if self.group:
                     fail("A 'defining_class' parameter cannot be in an optional group.")
+                if self.block.signatures[-1].cls is None:
+                    fail("A 'defining_class' parameter can be specified inside a class method.")
```
Comment (vstinner) [inline Tools/clinic/libclinic/dsl_parser.py]: I don't understand the error message. Do you mean "can only be defined"?

Maybe say "cannot be defined at module level: can only be defined on a class method"? Or something like that.

**Comment by neonene** [comment]
I cannot investigate the mypy error right now, failing to build mypy on 3.13.

**Comment by erlend-aasland** [comment]
> I cannot investigate the mypy error right now, failing to build mypy on 3.13.

I can take a look. Would you like me to update the PR for you?

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by neonene** [comment]
Thank you. I just managed to install mypy. I'll try it a bit more.

**Comment by neonene** [comment]
Please check.

File: Tools/clinic/libclinic/dsl_parser.py [hunk 2]
```diff
@@ -1102,6 +1102,9 @@ def bad_node(self, node: ast.AST) -> None:
                     fail("A 'defining_class' parameter cannot have a default value.")
                 if self.group:
                     fail("A 'defining_class' parameter cannot be in an optional group.")
+                assert isinstance(self.block.signatures[-1], Function)
+                if self.block.signatures[-1].cls is None:
```
Comment (erlend-aasland) [inline Tools/clinic/libclinic/dsl_parser.py]: Can't you just check `self.function.cls`?

Comment (neonene) [reply Tools/clinic/libclinic/dsl_parser.py]: Applied. There was no need to check the mypy.  Am I missing something else?

Comment (erlend-aasland) [reply Tools/clinic/libclinic/dsl_parser.py]: Yes, mypy should be fine with the `self.function.cls` check. Thanks!

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thank you, @neonene and @vstinner!

**Pull request merged**